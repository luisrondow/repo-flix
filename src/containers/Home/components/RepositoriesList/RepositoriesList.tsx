import Text from '../../../../components/Text'

import RepositoryCard from '../RepositoryCard'
import { RepositoriesListProps } from './RepositoriesList.types'
import ArrowIcon from '../../../../components/Icons/ArrowIcon'
import useRepositories from '../../../../hooks/useRepositories'
import useHorizontalScroll from '../../../../hooks/useHorizontalScroll'

const RepositoriesList = (props: RepositoriesListProps) => {
  const { title, techName } = props

  const { repositories, isLoading, isError, error } = useRepositories(techName, 'stars')
  const { listRef, isListLeftScrolled, handleScroll } = useHorizontalScroll()

  if (isLoading) return <div>Loading...</div>
  if (isError) {
    return <pre>{JSON.stringify(error)}</pre>
  }

  return (
    <div data-testid="repositories-list-container" className="relative">
      <Text as="h1" className="pl-16">
        {title}
      </Text>
      <div
        data-testid="repositories-list"
        ref={listRef}
        className="no-scrollbar relative flex h-72 w-full snap-mandatory flex-row flex-nowrap items-center space-x-4 overflow-x-auto overflow-y-hidden py-4 pl-16 pr-4"
      >
        {repositories?.map((repository) => {
          const { id, name, url, image } = repository

          return (
            <RepositoryCard
              key={id}
              type="list"
              repository={{ id, name, image }}
              onRepositoryClick={() => window.open(url, '_blank')}
            />
          )
        })}
      </div>
      <div className="absolute bottom-0 right-0 z-20 mb-4 flex h-64 w-60 items-center justify-end bg-gradient-to-r from-transparent to-base opacity-0 transition-opacity hover:opacity-100">
        <button onClick={() => handleScroll('right')} className="h-full pl-8 pr-2">
          <ArrowIcon direction="right" />
        </button>
      </div>
      {isListLeftScrolled ? (
        <div className="absolute bottom-0 left-0 z-20 mb-4 flex h-64 w-60 items-center justify-start bg-gradient-to-l from-transparent to-base opacity-0 transition-opacity hover:opacity-100">
          <button onClick={() => handleScroll('left')} className="h-full pr-8 pl-2">
            <ArrowIcon direction="left" />
          </button>
        </div>
      ) : null}
    </div>
  )
}

export default RepositoriesList
