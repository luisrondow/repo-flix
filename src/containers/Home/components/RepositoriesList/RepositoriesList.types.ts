export type RepositoriesListProps = {
  title: string
  techName: string
}
