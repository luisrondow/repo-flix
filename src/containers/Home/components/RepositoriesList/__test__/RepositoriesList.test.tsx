import { act, render } from '@testing-library/react'
import RepositoriesList from '../RepositoriesList'

const repositories = [
  {
    id: '10270250',
    name: 'facebook/react',
    url: 'https://github.com/facebook/react',
    image: 'https://opengraph.githubassets.com/123abc/facebook/react',
  },
  {
    id: '126577260',
    name: 'trekhleb/javascript-algorithms',
    url: 'https://github.com/trekhleb/javascript-algorithms',
    image: 'https://opengraph.githubassets.com/123abc/trekhleb/javascript-algorithms',
  },
  {
    id: '2126244',
    name: 'twbs/bootstrap',
    url: 'https://github.com/twbs/bootstrap',
    image: 'https://opengraph.githubassets.com/123abc/twbs/bootstrap',
  },
  {
    id: '6498492',
    name: 'airbnb/javascript',
    url: 'https://github.com/airbnb/javascript',
    image: 'https://opengraph.githubassets.com/123abc/airbnb/javascript',
  },
  {
    id: '29028775',
    name: 'facebook/react-native',
    url: 'https://github.com/facebook/react-native',
    image: 'https://opengraph.githubassets.com/123abc/facebook/react-native',
  },
]

jest.mock('../../../../../hooks/useRepositories', () => {
  return {
    __esModule: true,
    default: () => ({
      repositories,
      isLoading: false,
      isError: false,
      error: null,
    }),
  }
})

describe('RepositoriesList', () => {
  it('should render the correctly', () => {
    const { container, getByTestId, getByText, getAllByRole } = render(
      <RepositoriesList title="Mock title" techName="Mock" />,
    )

    const ids = repositories.map((repository) => repository.id)

    expect(getByTestId('repositories-list-container')).toBeInTheDocument()
    expect(getByTestId('repositories-list')).toBeInTheDocument()
    expect(getByText('Mock title')).toBeInTheDocument()
    expect(getAllByRole('button')).toHaveLength(1)

    ids.forEach((id) => {
      expect(getByTestId(`repository-card-${id}`)).toBeInTheDocument()
    })

    expect(container).toMatchSnapshot()
  })

  it('should scroll the list correctly', () => {
    const { getByTestId, getAllByRole } = render(
      <RepositoriesList title="Mock title" techName="Mock" />,
    )

    window.HTMLElement.prototype.scrollBy = jest.fn()

    const list = getByTestId('repositories-list')
    let buttons = getAllByRole('button')

    act(() => buttons[0].click())

    expect(list.scrollBy).toHaveBeenCalledWith({
      left: 448,
      behavior: 'smooth',
    })
    expect(getAllByRole('button')).toHaveLength(2)

    buttons = getAllByRole('button')

    act(() => buttons[1].click())

    expect(list.scrollBy).toHaveBeenCalledWith({
      left: -448,
      behavior: 'smooth',
    })
    expect(getAllByRole('button')).toHaveLength(1)
  })
})
