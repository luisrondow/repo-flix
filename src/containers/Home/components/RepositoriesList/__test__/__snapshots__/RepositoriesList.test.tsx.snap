// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RepositoriesList should render the correctly 1`] = `
<div>
  <div
    class="relative"
    data-testid="repositories-list-container"
  >
    <h1
      class="pl-16 text-[#e0def4] font-roboto text-4xl font-bold"
    >
      Mock title
    </h1>
    <div
      class="no-scrollbar relative flex h-72 w-full snap-mandatory flex-row flex-nowrap items-center space-x-4 overflow-x-auto overflow-y-hidden py-4 pl-16 pr-4"
      data-testid="repositories-list"
    >
      <div
        class="relative h-56 w-112 flex-shrink-0 flex-grow basis-112 cursor-pointer snap-center scroll-mx-96 transition-transform hover:z-10 hover:scale-115 hover:shadow-md"
        data-testid="repository-card-10270250"
      >
        <img
          alt="facebook/react"
          class="h-full w-full"
          src="https://opengraph.githubassets.com/123abc/facebook/react"
        />
      </div>
      <div
        class="relative h-56 w-112 flex-shrink-0 flex-grow basis-112 cursor-pointer snap-center scroll-mx-96 transition-transform hover:z-10 hover:scale-115 hover:shadow-md"
        data-testid="repository-card-126577260"
      >
        <img
          alt="trekhleb/javascript-algorithms"
          class="h-full w-full"
          src="https://opengraph.githubassets.com/123abc/trekhleb/javascript-algorithms"
        />
      </div>
      <div
        class="relative h-56 w-112 flex-shrink-0 flex-grow basis-112 cursor-pointer snap-center scroll-mx-96 transition-transform hover:z-10 hover:scale-115 hover:shadow-md"
        data-testid="repository-card-2126244"
      >
        <img
          alt="twbs/bootstrap"
          class="h-full w-full"
          src="https://opengraph.githubassets.com/123abc/twbs/bootstrap"
        />
      </div>
      <div
        class="relative h-56 w-112 flex-shrink-0 flex-grow basis-112 cursor-pointer snap-center scroll-mx-96 transition-transform hover:z-10 hover:scale-115 hover:shadow-md"
        data-testid="repository-card-6498492"
      >
        <img
          alt="airbnb/javascript"
          class="h-full w-full"
          src="https://opengraph.githubassets.com/123abc/airbnb/javascript"
        />
      </div>
      <div
        class="relative h-56 w-112 flex-shrink-0 flex-grow basis-112 cursor-pointer snap-center scroll-mx-96 transition-transform hover:z-10 hover:scale-115 hover:shadow-md"
        data-testid="repository-card-29028775"
      >
        <img
          alt="facebook/react-native"
          class="h-full w-full"
          src="https://opengraph.githubassets.com/123abc/facebook/react-native"
        />
      </div>
    </div>
    <div
      class="absolute bottom-0 right-0 z-20 mb-4 flex h-64 w-60 items-center justify-end bg-gradient-to-r from-transparent to-base opacity-0 transition-opacity hover:opacity-100"
    >
      <button
        class="h-full pl-8 pr-2"
      >
        <svg
          class="h-12 w-12 stroke-gray-light"
          fill="none"
          stroke="currentColor"
          stroke-width="1.5"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M8.25 4.5l7.5 7.5-7.5 7.5"
            stroke-linecap="round"
            stroke-linejoin="round"
          />
        </svg>
      </button>
    </div>
  </div>
</div>
`;
